File: src/example/basicGossip/BasicGossip.java
package example.basicGossip;

import peersim.cdsim.CDProtocol;
import peersim.config.Configuration;
import peersim.config.FastConfig;
import peersim.core.CommonState;
import peersim.core.Linkable;
import peersim.core.Node;
import peersim.edsim.EDProtocol;
import peersim.transport.Transport;
import peersim.vector.SingleValueHolder;

/**
 * Event driven version of epidemic averaging.
 */
public class BasicGossip extends SingleValueHolder implements CDProtocol,
		EDProtocol {

	public static int fanout;
	public static int networkSize;
	protected int info;

	// --------------------------------------------------------------------------
	// Initialization
	// --------------------------------------------------------------------------

	/**
	 * @param prefix
	 *            string prefix for config properties
	 */
	public BasicGossip(String prefix) {
		super(prefix);
		networkSize = (Configuration.getInt("network.size", 1));
		fanout = (int) Math.log(networkSize);
	}

	
	public int a(){
		return fanout;
	}

	// --------------------------------------------------------------------------
	// methods
	// --------------------------------------------------------------------------

	/**
	 * This is the standard method the define periodic activity. The frequency
	 * of execution of this method is defined by a
	 * {@link peersim.edsim.CDScheduler} component in the configuration.
	 */
	public void nextCycle(Node node, int pid) {
		// System.out.println("PID " + pid);
		// streamer
		if (node.getIndex() == 0) {
			System.out.println("streamer generating " + info);
			for (int i = 0; i < fanout; i++) {
				Linkable linkable = (Linkable) node.getProtocol(FastConfig
						.getLinkable(pid));
				if (linkable.degree() > 0) {
					Node peern = linkable.getNeighbor(CommonState.r
							.nextInt(linkable.degree()));

					// System.out.println("streamer sending to " +
					// peern.getIndex());

					// XXX quick and dirty handling of failures
					// (message would be lost anyway, we save time)
					if (!peern.isUp())
						return;

					System.out
							.println("streamer will send to " + peern.getID());

					((Transport) node.getProtocol(FastConfig.getTransport(pid)))
							.send(node, peern, new Info(info, node), pid);
				}

			}

			info++;
		}

	}

	// --------------------------------------------------------------------------

	/**
	 * This is the standard method to define to process incoming messages.
	 */
	public void processEvent(Node node, int pid, Object event) { /*
																 * 
																 * Info aem =
																 * (Info) event;
																 * //System.out.
																 * println
																 * ("Node: " +
																 * node
																 * .getIndex() +
																 * " received message "
																 * +
																 * aem.value());
																 * 
																 * 
																 * if
																 * (!((Usernode)
																 * node
																 * ).containsElem
																 * (
																 * aem.value()))
																 * { // first
																 * time we
																 * receive this
																 * message
																 * ((Usernode)
																 * node
																 * ).saveMessage
																 * (
																 * aem.value());
																 * for (int i =
																 * 0; i <
																 * fanout; i++)
																 * { Linkable
																 * linkable =
																 * (Linkable)
																 * node
																 * .getProtocol
																 * (FastConfig
																 * .getLinkable
																 * (pid)); if
																 * (linkable
																 * .degree() >
																 * 0) { Node
																 * peern =
																 * linkable
																 * .getNeighbor
																 * (CommonState
																 * .r
																 * .nextInt(linkable
																 * .degree()));
																 * 
																 * // XXX quick
																 * and dirty
																 * handling of
																 * failures //
																 * (message
																 * would be lost
																 * anyway, we
																 * save time) if
																 * (
																 * !peern.isUp()
																 * ) return;
																 * 
																 * ((Transport)
																 * node
																 * .getProtocol
																 * (FastConfig
																 * .getTransport
																 * (pid)))
																 * .send(node,
																 * peern, new
																 * Info(info,
																 * node), pid);
																 * }
																 * 
																 * }
																 * //System.out
																 * .println
																 * ("Node: " +
																 * node
																 * .getIndex()
																 * );
																 * //((Usernode)
																 * node
																 * ).dumpMessageList
																 * ();
																 * 
																 * }
																 */

		/*
		 * if (aem.sender != null) ((Transport)
		 * node.getProtocol(FastConfig.getTransport(pid))).send( node,
		 * aem.sender, new AverageMessage(value, null), pid);
		 * 
		 * value = (value + aem.value) / 2;
		 */
	}
}

File: src/basicGossip/protocols/BasicGossip.scala
package basicGossip.protocols

import peersim.cdsim.CDProtocol
import peersim.config.Configuration
import peersim.core.Linkable
import peersim.core.Network
import peersim.core.Node
import peersim.edsim.EDProtocol
import peersim.transport.Transport
import peersim.vector.SingleValueHolder
import basicGossip.messages.Info
import basicGossip.node.Usernode
import peersim.config.FastConfig
import scala.util.Random
import utils.DistinctRandom
import basicGossip.oracle.Oracle
import basicGossip.observers.AvgReliability
import basicGossip.protocols.GeneralProtocol.Heavyweight

class BasicGossip(prefix: String) extends SingleValueHolder(prefix) with CDProtocol with EDProtocol {

  val networkSize = (Configuration.getInt("network.size", 1))
  val cycles = (Configuration.getInt("CYCLES", 1))
  var info: Int = 0

  override def nextCycle(node: Node, pid: Int): Unit = {

    if (Network.size < 100) {
      Oracle.addAltruisticNode
    } else if (Oracle.currentPackage == 100) {
      Oracle.allNodesExceptStreamer map {
        node =>
          node.behaviorProtocol match {
            case h: Heavyweight => h.setManagers(node)
            case _ =>
          }
      }
    }

    if (Oracle.currentPackage == 999) {
      Oracle.injectFreeRiders
    }

    for (id <- 1 until Network.size) {
      val node = Oracle.getNode(id)
      node.getProtocol(3) match {
        case p: CDProtocol =>
          p.nextCycle(node, 3)
      }
      node.getProtocol(2) match {
        case p: CDProtocol =>
          p.nextCycle(node, 2)
      }
      //node.requestProposals

    }

    Oracle.updateCurrentPackage
    sendInfo(Oracle.getNode(0), newInfo, pid)
  }

  private def sendInfo(streamer: Usernode, info: Info, pid: Int) = {
    val link = Oracle.getLinkable(streamer)
    val trans = streamer.getProtocol(FastConfig.getTransport(pid))

    trans match {
      case t: Transport =>
        DistinctRandom.sample(1 until Network.size toList, Math.min(Oracle.fanout, Network.size)) map {
          //Random.shuffle(Oracle.altruistics).take(fanout) map {  
          id =>
            val peern = Oracle.getNode(id)
            if (peern.isUp) {
              //println("sendin to " + peern.getID)
              t.send(streamer, peern, info, pid)
            }
        }
    }
  }

  private def newInfo: Info = {
    val currentInfo = info;
    info += 1;
    Info(currentInfo, Oracle.getNode(0), 0)
  }

  def processEvent(node: Node, pid: Int, event: Object) {
  }

}

object BasicGossip extends BasicGossip("Basic Gossip")